lib/mp-readline/readline.c

```diff
@@ -98,6 +98,72 @@ typedef struct _readline_t {
 
 STATIC readline_t rl;
 
+#if MICROPY_REPL_EMACS_WORDS_MOVE
+STATIC size_t cursor_count_forward_word(size_t cursor_pos, vstr_t *line) {
+    size_t pos = rl.cursor_pos;
+    const char *line_buf = vstr_str(line);
+
+    if (pos == vstr_len(rl.line)) {
+        return pos;
+    }
+
+    assert(pos < vstr_len(rl.line));
+    if (!unichar_isalnum(line_buf[pos])) {
+        pos++;
+
+        while (pos < vstr_len(rl.line)) {
+            if (unichar_isalnum(line_buf[pos])) {
+                break;
+            }
+            pos++;
+        }
+    }
+
+    if (pos == rl.line->len) {
+        return pos;
+    }
+
+    pos++;
+    while (pos < rl.line->len) {
+        if (!unichar_isalnum(line_buf[pos])) {
+            break;
+        }
+        pos++;
+    }
+
+    return pos - rl.cursor_pos;
+}
+
+STATIC size_t cursor_count_backward_word(size_t cursor_pos, vstr_t *line) {
+    size_t pos = rl.cursor_pos;
+    const char *line_buf = vstr_str(line);
+
+    if (pos == 0) {
+        return pos;
+    }
+
+    pos--; // this also ensures pos <= line length
```

Comment:
> this also ensures pos <= line length

I assume you added this comment to explain why you don't use an `assert` like above. Things is, the comment is only correct under the assumption that cursor pos never crosses line length. Which it shouldn't do obviously. But that's actually another reason to use the assert: make sure the calling code behaves, makes sure this code doesn't access out of bounds data, and a correct way of commenting expected behaviour.

Reply:
Actually, without this decrement you'd have to check it in runtime. So by this comment I wanted to remind the reader why we don't need to `if` it here.

But you are absolutely right about the use of asserts. I'll add one here as well.